### ISSUE TITLE
Recommendation for healthcheck

### ISSUE THREAD

**Issue by ericonukign** [ruby-shoryuken/shoryuken#674]
Hi,

Scenario: Shoryuken running in a kubernetes pod.

Recently I came up with a problem: I need to perform a healthcheck on a pod that is running shoryuken. Since shoryuken does not start a webserver, how can I perform a healthcheck on the pod (and the workers)? I've already seen the kill -TTIN command, and came up with a scenario where a script would read the shoryuken output and return either 0 if it finds the threads running in the pod log output. But it seems a bit clunky, I was wondering if there is a more straightforward solution for this.

Thank you for your time.

**Comment by cjlarose** [comment]
There's not currently a good health check mechanism for shoryuken that I know about. Checking to see if the process is running of course isn't sufficient since in theory it could be locked up in a way that it isn't making progress on jobs.

The analog for a web server health check would be to request the local shoryuken instance to perform some sort of "health check" job that goes through the whole system, but there doesn't exist a way to force shoryuken to pick up any work except for what's in the SQS queues.

I thought sidekiq might be a source of inspiration here since it's a similar program in that sense, but it looks to me that they don't have a good answer here either: https://github.com/mperham/sidekiq/issues/2936

**Comment by dudo** [comment]
What is you had a PingWorker that took a message from SQS, spit out some logs / emitted some metrics, then scheduled another job to run in 1(N) minute?

There are also better monitoring tools for sparse work. Cronitor comes to mind. 

Regardless, I don't think this is a role that Shoryuken needs to support. I'm happy to riff on more ideas, though!

**Comment by ericonukign** [comment]
Thank you all for the comments, still looking for a solution though

**Comment by cjlarose** [comment]
> What if you had a PingWorker that took a message from SQS, spit out some logs / emitted some metrics, then scheduled another job to run in 1(N) minute?

My company does exactly this in production. This kind of approach is good for monitoring that SQS messages are being processed across the system (across potentially several instances), which is usually good enough. The kubernetes health check is more a of check of whether or not a _particular_ instance is able to make progress consuming work. There's not a good existing mechanism for this now.

To play devil's advocate though, having a per-instance health check isn't as important for a background instance as it is for a web server since if the instance is not able to pick up work, other instances will pull the work from the queue. If it is able to pick up work, but cannot complete the jobs, the visibility timeout should expire and the work will be completed by another worker. In other words, there are other mechanisms that can build confidence and resilience against failures besides per-instance health checks.

**Comment by cjlarose** [comment]
> I’m not sure what that webserver should actually check or whether it would be better to start it up by forking or by starting a new thread.

I think this is the real root of the problem: even if shoryuken _did_ provide some sort of interface for performing a health check, it's not really clear what it would do because it's difficult to test the whole system end-to-end (with a roundtrip to SQS).

I think the best thing we could probably do is to `post` a job into each of the `Manager`s' thread pool executors and wait until the jobs successfully complete. I'm open to a PR that adds a health check that does exactly this.

I'm open to adding an HTTP server (or just raw TCP server) to shoryuken on an opt-in basis (similar to puma's [control/status server](https://github.com/puma/puma/blob/61dd7f4670da7ba6895fe7292340cabe90fc5bff/README.md#controlstatus-server)). I'm hesitant to communicate that server's API would be "public API" though because it makes it harder for us to make changes to it. Instead, tools like k8s allow you to just specify a command that must return a non-zero exit code to indicate that the process is unhealthy. I think we can probably just add something to the shoryuken CLI like `bundle exec shoryuken healthcheck --socket tcp://localhost:9999`. That process would then use the internal HTTP or TCP server for doing the health check itself, but then just exit with the appropriate code.

**Comment by LyleDavis** [comment]
> > I’m not sure what that webserver should actually check or whether it would be better to start it up by forking or by starting a new thread.
> 
> I think this is the real root of the problem: even if shoryuken _did_ provide some sort of interface for performing a health check, it's not really clear what it would do because it's difficult to test the whole system end-to-end (with a roundtrip to SQS).
> I think the best thing we could probably do is to post a job into each of the Managers' thread pool executors and wait until the jobs successfully complete. I'm open to a PR that adds a health check that does exactly this.

@cjlarose Does it really need to check that they complete, or does it just need to know that a threadpool executor is running?

I think it's worth clarifying that a worker is fundamentally different to a webserver, and there are two very different kinds of health checks that should be performed on a worker. They work in tandem because they give different things.

The first is a simple "are you up?" check for the worker process itself. This is the kind of thing that would be used to terminate bad containers where the shoryuken process died or never started before messages begin to build up - to keep your fleet at max capacity.

The second is monitoring of the queues themselves, either via cloudwatch or something else. That is what actually tells you the health of your processing pipeline, whether you need to scale out or in, whether you're failing too many messages, etc.

Maybe it could be as simple as a webserver with a health endpoint that, when hit, checks that all the shoryuken groups have a manager and that each manager responds with `true` to `#running?`? https://github.com/ruby-shoryuken/shoryuken/blob/ce1ceb0b827cbd2fc8e7e95dee7da0238754d985/lib/shoryuken/manager.rb#L27

**Comment by cjlarose** [comment]
> I think it's worth clarifying that a worker is fundamentally different to a webserver, and there are two very different kinds of health checks that should be performed on a worker. They work in tandem because they give different things.

I disagree with this sentiment. Obviously webservers and background job processing systems are different and have different use-cases, but webservers often have a very similar architecture to a background job processor: there's a queue at the front (for a webserver, it's often a TCP socket), and a collection of workers that process the messages from that queue. For any request-processing system, a reasonable health check would be to try to exercise as much of that pipeline as possible. For a webserver, that often means actually executing an HTTP request, since the expectation is that it'll be picked up from a queue and executed by a worker like any other legitimate request. For shoryuken, the best possible health check would be to send actual messages to SQS such that they'd be picked up by the particular worker you're trying to test. That's a bit cumbersome of course because unless you have separate SQS queues per worker, it's not really practical.



> Maybe it could be as simple as a webserver with a configurable health endpoint that, when hit, checks that all the shoryuken groups have a manager and that each manager responds with `true` to `#running?`?

I'm open to this as an MVP. I'm a little hesitant that this would give some false positives, but it's better than nothing. I don't think it necessarily has to be configurable: the `Manager#running?` check seems fine.



> Or alternatively, Shoryuken expose that check as a public API and let consumers start up their own webserver if they want during the `on_start` callback. They would be responsible for terminating that thread during `on_stop` as well if they were to do that.

I really like this idea. I think `Manager` instances are already available from the `Shoryuken` module object, but I'm not 100% sure. But if that's true, it would mean that healthcheck functionality could be offered by a gem. The gem would add all of the appropriate `on_start` and `on_stop` callbacks to spin up a webserver. That would be great for users that don't want to have to pay the cost of the healthcheck functionality in `shoryuken` itself (especially if it would mean we'd have to add more dependencies).

**Comment by tibbon** [comment]
My other idea is to simply have the startup callback create a file, and my liveness probe checks for the presence of that file.
